lib/shared/puppet_forge/unpacker.rb

```diff
@@ -0,0 +1,78 @@
+require 'pathname'
+require 'tmpdir'
+require 'shared/puppet_forge/error'
+require 'shared/puppet_forge/tar'
+
+module PuppetForge
+  class Unpacker
+    # Unpack a tar file into a specified directory
+    #
+    # @param filename [String] the file to unpack
+    # @param target [String] the target directory to unpack into
+    def self.unpack(filename, target, tmpdir)
+      inst = self.new(filename, target, tmpdir)
+      inst.unpack
+      inst.move_into(target)
+    end
+
+    # Set the owner/group of the target directory to those of the source
+    #
+    # @param source [Pathname] source of the permissions
+    # @param target [Pathname] target of the permissions change
+    # @param microsoft_windows boolean flag for whether this system is windows or not
+    def self.harmonize_ownership(source, target, microsoft_windows = false)
+      if !microsoft_windows
+        FileUtils.chown_R(source.stat.uid, source.stat.gid, target)
+      end
+    end
+
+    # @param filename [String] the file to unpack
+    # @param target [String] the target directory to unpack into
+    def initialize(filename, target, tmpdir)
+      @filename = Pathname.new(filename)
+      @target = target
+      @tmpdir = tmpdir
+    end
+
+    # @api private
+    def unpack
+      begin
+        PuppetForge::Tar.instance.unpack(@filename.to_s, @tmpdir)
+      rescue PuppetForge::ExecutionFailure => e
+        raise RuntimeError, "Could not extract contents of module archive: #{e.message}"
+      end
+    end
+
+    # @api private
+    def move_into(dir)
```

Comment:
The behavior of FileUtils::mv is correct for the copying of the contents of the root_dir to the dir, so this doesn't need to change.